### PR TITLE
Adding r8brain-free-src

### PR DESCRIPTION
**Pull request by johanneskares**
Specify library name and version:  **r8brain-free-src/4.6**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/r8brain-free-src/all/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project(r8brain-free-src CXX)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_library(r8brain source_subfolder/r8bbase.cpp)
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/CMakeLists.txt:7]:
```suggestion
add_library(r8brain source_subfolder/r8bbase.cpp)
if(MSVC AND BUILD_SHARED_LIBS)
  set_property(TARGET r8brain PROPERTY WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

install(
  TARGETS r8brain
  RUNTIME DESTINATION bin
  LIBRARY DESTINATION lib
  ARCHIVE DESTINATION lib
)

file(GLOB PUBLIC_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/source_subfolder/*.h)
install(FILES ${PUBLIC_HEADERS} DESTINATION include)
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    name = "r8brain-free-src"
+    version = "4.6"
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/avaneev/r8brain-free-srcn"
+    description = "https://github.com/avaneev/r8brain-free-src"
+    topics = ("Audio", "Sample Rate", "Conversion")
+    exports_sources = ["CMakeLists.txt"]
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:16]:
```suggestion
    generators = "cmake"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": False, "fPIC": True}
    
    _cmake = None
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import os
+
+from conans import ConanFile, CMake, tools
+
+
+class R8brainFreeSrcConan(ConanFile):
+    name = "r8brain-free-src"
+    version = "4.6"
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:8]:
```suggestion
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/avaneev/r8brain-free-srcn"
+    description = "https://github.com/avaneev/r8brain-free-src"
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:12]:
```suggestion
    description = "High-quality pro audio sample rate converter / resampler C++ library"
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        cmake.build()
+
+    def package(self):
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("*.h", dst="include", src=self._source_subfolder)
+        self.copy("*.lib", dst="lib", keep_path=False)
+        self.copy("*.dll", dst="bin", keep_path=False)
+        self.copy("*.so", dst="lib", keep_path=False)
+        self.copy("*.dylib", dst="lib", keep_path=False)
+        self.copy("*.a", dst="lib", keep_path=False)
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:39]:
```suggestion
    def _configure_cmake(self):
        if self._cmake:
            return self._cmake
        self._cmake = CMake(self)
        self._cmake.configure()
        return self._cmake

    def build(self):
        cmake = self._configure_cmake()
        cmake.build()

    def package(self):
        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
        cmake = self._configure_cmake()
        cmake.install()
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    description = "https://github.com/avaneev/r8brain-free-src"
+    topics = ("Audio", "Sample Rate", "Conversion")
+    exports_sources = ["CMakeLists.txt"]
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:21]:
```suggestion

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            del self.options.fPIC

```

**Comment by johanneskares** [comment]
Thanks @SpaceIm! Very helpful!

**Comment by johanneskares** [comment]
@SpaceIm 

`WINDOWS_EXPORT_ALL_SYMBOLS` doesn't work, because that feature can't handle the static global data of this file:

https://github.com/avaneev/r8brain-free-src/blob/master/r8bbase.cpp

as mentioned in the CMake docs:

https://cmake.org/cmake/help/v3.18/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html

So I guess for now we can't support Windows Shared Libs without changes to the source code, correct?

**Comment by SpaceIm** [comment]
Ok I see. Do not delete shared option but instead, in configure method, raise ConanInvalidConfiguration if Visual Studio and shared.

File: recipes/r8brain-free-src/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+            del self.options.shared
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:28]:
```suggestion
```

File: recipes/r8brain-free-src/all/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+if(MSVC AND BUILD_SHARED_LIBS)
+  set_property(TARGET r8brain PROPERTY WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+endif()
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/CMakeLists.txt:10]:
```suggestion
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
+import os
+
+from conans import ConanFile, CMake, tools
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:3]:
```suggestion
from conans import ConanFile, CMake, tools
from conans.errors import ConanInvalidConfiguration
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+            del self.options.shared
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
```
Comment (SpaceIm) [inline recipes/r8brain-free-src/all/conanfile.py:32]:
```suggestion
            del self.options.fPIC
        if self.settings.compiler == "Visual Studio" and self.options.shared:
            raise ConanInvalidConfiguration("Shared r8brain-free-src cannot be built with Visual Studio")
```

File: recipes/r8brain-free-src/all/conandata.yml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "4.6":
+    sha256: 9a0d66c3a04954359ef9f69168b58fcdf11e8975821181ee7d1e3b42e8a2cf0e
+    url: http://github.com/avaneev/r8brain-free-src/archive/version-4.6.tar.gz
```
Comment (uilianries) [inline recipes/r8brain-free-src/all/conandata.yml:4]:
```suggestion
    url: https://github.com/avaneev/r8brain-free-src/archive/version-4.6.tar.gz
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class R8brainFreeSrcConan(ConanFile):
+    name = "r8brain-free-src"
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/avaneev/r8brain-free-srcn"
```
Comment (uilianries) [inline recipes/r8brain-free-src/all/conanfile.py:11]:
```suggestion
    homepage = "https://github.com/avaneev/r8brain-free-src"
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    description = "High-quality pro audio sample rate converter / resampler C++ library"
+    topics = ("Audio", "Sample Rate", "Conversion")
```
Comment (uilianries) [inline recipes/r8brain-free-src/all/conanfile.py:13]:
```suggestion
    topics = ("audio", "sample-rate", "conversion", "audio-processing", "resampler")
```

File: recipes/r8brain-free-src/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        if self.settings.compiler == "Visual Studio" and self.options.shared:
+            raise ConanInvalidConfiguration("Shared r8brain-free-src cannot be built with Visual Studio")
```
Comment (madebr) [inline recipes/r8brain-free-src/all/conanfile.py:34]: Adding [`CMAKE_EXPORT_WINDOWS_SYMBOLS`](https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) doesn't work?

Comment (johanneskares) [reply recipes/r8brain-free-src/all/conanfile.py:34]: see above the comments:

[Click](https://github.com/conan-io/conan-center-index/pull/2536#issuecomment-674457138)

**Review by SpaceIm — APPROVED** [review]

**Pull request merged**